consensus/tendermint/messages.go

```diff
@@ -0,0 +1,107 @@
+package tendermint
+
+// Todo: Signature over the messages needs to be handled somewhere. There are 2 options:
+//	1. Add the signature to each message and extend the Validator Set interface to include VerifyMessageSignature
+//	method.
+//	2. The P2P layer signs the message before gossiping to other validators and verifies the signature before passing
+//	the message to the consensus engine.
+//  The benefit of P2P layer handling the verification of the signature is the that the consensus layer can assume
+//  the message is from a validator in the validator set. However, this means that the P2P layer would need to be aware
+//  of the validator set and would need access to the blockchain which may not be a good idea.
+
+type Message[V Hashable[H], H Hash, A Addr] interface {
+	Proposal[V, H, A] | Prevote[H, A] | Precommit[H, A]
+}
+
+type Proposal[V Hashable[H], H Hash, A Addr] struct {
+	H          height
+	R          round
+	ValidRound int
+	Value      *V
+
+	Sender A
+}
+
+type (
+	Prevote[H Hash, A Addr]   Vote[H, A]
+	Precommit[H Hash, A Addr] Vote[H, A]
+)
+
+type Vote[H Hash, A Addr] struct {
+	H  height
+	R  round
+	ID *H
+
+	Sender A
+}
+
+// messages keep tracks of all the proposals, prevotes, precommits by creating a map structure as follows:
+// height->round->address->[]Message
+
+// Todo: would the following representation of message be better:
+//
+//	  height -> round -> address -> ID -> Message
+//	How would we keep track of nil votes? In golan map key cannot be nil.
+//	It is not easy to calculate a zero value when dealing with generics.
```

Comment:
I think it's possible with
```
func ZeroValue[T any]() T {
	var zero T
	return zero
}
```
Although it's not necessary a good way, depending whether zero value is not a valid value.

Reply:
That is a possible solution but I am not sure whether it would work